lib_c/dates_calc.c

```diff
@@ -0,0 +1,340 @@
+}
+
+void dc_sub_periods (dc_period *ret, const dc_period *p1, const dc_period *p2) {
+  ret->years = p1->years - p2->years;
+  ret->months = p1->months - p2->months;
+  ret->days = p1->days - p2->days;
+}
+
+void dc_mul_periods (dc_period *ret, const dc_period *p, const long int m) {
+  ret->years = p->years * m;
+  ret->months = p->months * m;
+  ret->days = p->days * m;
+}
+
+dc_success dc_period_to_days (long int *ret, const dc_period *p) {
+  if (p->years || p->months) return dc_error;
+  else *ret = p->days;
+  return dc_ok;
+}
+
+BOOL dc_is_leap_year (const long int y) {
+  return (y % 400 == 0 || (y % 4 == 0 && y % 100 != 0));
+}
+
+unsigned long int dc_days_in_month (const dc_date *d) {
+  switch (d->month) {
+  case 2:
+    return (dc_is_leap_year(d->year) ? 29 : 28);
+  case 4: case 6: case 9: case 11:
+    return 30;
+  default:
+    return 31;
+  }
+}
+
+BOOL dc_is_valid_date (const dc_date *d) {
+  return (1 <= d->day && d->day <= dc_days_in_month(d));
+}
+
+dc_success dc_make_date(dc_date *ret, const long int y, const unsigned long int m, const unsigned long int d) {
+  ret->year = y;
+  ret->month = m;
+  ret->day = d;
+  if (dc_is_valid_date(ret)) return dc_ok;
+  else return dc_error;
+}
+
+void dc_copy_date(dc_date *ret, const dc_date *d) {
+  if (ret != d) {
+    ret->year = d->year;
+    ret->month = d->month;
+    ret->day = d->day;
+  }
+}
+
+/* Precondition: [1 <= d->month <= 12]. The returned day is always [1] */
+void dc_add_months(dc_date *ret, const dc_date *d, const long int months) {
+  long int month = d->month + months;
+  ret->day = 1;
+  ret->month = month >= 0 ? month % 12 : month % 12 + 12;
+  ret->year = d->year + (month >= 0 ? month / 12 : month / 12 - 1);
+}
+
+/* If the date is valid, does nothing. We expect the month number to be always
+   valid when calling this. If the date is invalid due to the day number, then
+   this function rounds down: if the day number is >= days_in_month, to the last
+   day of the current month. */
+void dc_prev_valid_date (dc_date *ret, const dc_date *d) {
+  assert (1 <= d->month && d->month <= 12);
+  assert (1 <= d->day && d->day <= 31);
+  if (dc_is_valid_date(d))
+    dc_copy_date(ret, d);
+  else {
+    ret->year = d->year;
+    ret->month = d->month;
+    ret->day = dc_days_in_month(d);
+  }
+}
+
+/* If the date is valid, does nothing. We expect the month number to be always
+   valid when calling this. If the date is invalid due to the day number, then
+   this function rounds down: if the day number is >= days_in_month, to the
+   first day of the next month. */
+void dc_next_valid_date (dc_date *ret, const dc_date *d) {
+  assert (1 <= d->month && d->month <= 12);
+  assert (1 <= d->day && d->day <= 31);
+  if (dc_is_valid_date(d))
+    dc_copy_date(ret, d);
+  else {
+    dc_add_months (ret, d, 1);
+  }
+}
+
+dc_success dc_round_date (dc_date *ret, const dc_date_rounding rnd, const dc_date *d) {
+  if (dc_is_valid_date(d)) {
+    *ret = *d;
```

Comment:
It seems you used copy in the other cases, why not here?

Reply:
Indeed, no good reason not to.